packages/ai/src/ui/chat.ts

```diff
@@ -93,11 +94,10 @@ export interface ChatState<UI_MESSAGE extends UIMessage> {
 
 export type ChatOnErrorCallback = (error: Error) => void;
 
-export type ChatOnToolCallCallback = ({
-  toolCall,
-}: {
-  toolCall: ToolCall<string, unknown>;
-}) => void | Promise<unknown> | unknown;
+export type ChatOnToolCallCallback<UI_MESSAGE extends UIMessage = UIMessage> =
+  (options: {
+    toolCall: InferUIMessageToolCall<UI_MESSAGE>;
+  }) => void | Promise<unknown> | unknown;
```

Comment:
is there an easy way to type the output as well, i.e. replace unknown here?

Reply:
@lgrammel I did a bit of experimenting yesterday, and it seems the easiest approach is to use a union type of all possible outputs as the return type. I was hoping TypeScript would support something more precise where a specific tool branch could enforce a matching return type, but that doesn't seem to be possible at the moment. I've pushed the change with the union type, please let me know if that's all right or whether I should revert it.